File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/buttons/ControlDrawer.java
package net.kdt.pojavlaunch.customcontrols.buttons;

import android.annotation.SuppressLint;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import net.kdt.pojavlaunch.customcontrols.ControlData;
import net.kdt.pojavlaunch.customcontrols.ControlDrawerData;
import net.kdt.pojavlaunch.customcontrols.ControlLayout;
import net.kdt.pojavlaunch.customcontrols.handleview.EditControlPopup;

import java.util.ArrayList;



@SuppressLint("ViewConstructor")
public class ControlDrawer extends ControlButton {


    public final ArrayList<ControlSubButton> buttons;
    public final ControlDrawerData drawerData;
    public final ControlLayout parentLayout;
    public boolean areButtonsVisible;
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public ControlDrawer(ControlLayout layout, ControlDrawerData drawerData) {
        super(layout, drawerData.properties);

        buttons = new ArrayList<>(drawerData.buttonProperties.size());
        this.parentLayout = layout;
        this.drawerData = drawerData;
        areButtonsVisible = layout.getModifiable();
    }
/**
 * 「add Button」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public void addButton(ControlData properties){
        addButton(new ControlSubButton(parentLayout, properties, this));
    }
/**
 * 「add Button」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public void addButton(ControlSubButton button){
        buttons.add(button);
        syncButtons();
        setControlButtonVisibility(button, areButtonsVisible);
    }
/**
 * 「ControlButtonVisibility」の値を設定します。
 */
    private void setControlButtonVisibility(ControlButton button, boolean isVisible){
        button.getControlView().setVisibility(isVisible ? VISIBLE : GONE);
    }
/**
 * 「switch Button Visibility」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private void switchButtonVisibility(){
        areButtonsVisible = !areButtonsVisible;
        int visibility = areButtonsVisible ? VISIBLE : GONE;
        for(ControlButton button : buttons){
            button.getControlView().setVisibility(visibility);
        }
    }
/**
 * 「align Buttons」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */

    //同期処理
    private void alignButtons(){
        if(buttons == null) return;
        if(drawerData.orientation == ControlDrawerData.Orientation.FREE) return;
        int margin = (int) ControlInterface.getMarginDistance();

        for(int i = 0; i < buttons.size(); ++i){
            switch (drawerData.orientation){
                case RIGHT:
                    buttons.get(i).setDynamicX(generateDynamicX(getX() + (drawerData.properties.getWidth() + margin)*(i+1) ));
                    buttons.get(i).setDynamicY(generateDynamicY(getY()));
                    break;

                case LEFT:
                    buttons.get(i).setDynamicX(generateDynamicX(getX() - (drawerData.properties.getWidth() + margin)*(i+1)));
                    buttons.get(i).setDynamicY(generateDynamicY(getY()));
                    break;

                case UP:
                    buttons.get(i).setDynamicY(generateDynamicY(getY() - (drawerData.properties.getHeight() + margin)*(i+1)));
                    buttons.get(i).setDynamicX(generateDynamicX(getX()));
                    break;

                case DOWN:
                    buttons.get(i).setDynamicY(generateDynamicY(getY() + (drawerData.properties.getHeight() + margin)*(i+1)));
                    buttons.get(i).setDynamicX(generateDynamicX(getX()));
                    break;
            }
            buttons.get(i).updateProperties();
        }
    }
/**
 * 「resize Buttons」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */

    private void resizeButtons(){
        if (buttons == null || drawerData.orientation == ControlDrawerData.Orientation.FREE) return;
        for(ControlSubButton subButton : buttons){
            subButton.mProperties.setWidth(mProperties.getWidth());
            subButton.mProperties.setHeight(mProperties.getHeight());

            subButton.updateProperties();
        }
    }
/**
 * 「sync Buttons」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public void syncButtons(){
        alignButtons();
        resizeButtons();
    }

    /**
     * パラメータとして渡されたボタンがこのドロワーに属するかどうかを確認します。
     *
     * @param button 検索するボタン
     * @return ボタンがドロワーのボタンリスト内にあるかどうか
     */
    public boolean containsChild(ControlInterface button){
        for(ControlButton childButton : buttons){
            if (childButton == button) return true;
        }
        return false;
    }
/**
 * 「pre Process Properties」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public ControlData preProcessProperties(ControlData properties, ControlLayout layout) {
        ControlData data = super.preProcessProperties(properties, layout);
        data.isHideable = true;
        return data;
    }
/**
 * 「Visible」の値を設定します。
 */
    @Override
    public void setVisible(boolean isVisible) {
        int visibility = isVisible ? VISIBLE : GONE;
        setVisibility(visibility);
        if(visibility == GONE || areButtonsVisible) {
            for(ControlSubButton button : buttons){
                button.getControlView().setVisibility(isVisible ? VISIBLE : (!mProperties.isHideable && getVisibility() == GONE) ? VISIBLE : View.GONE);
            }
        }
    }
/**
 * タッチイベントを処理します。
 * ユーザーからのタッチ入力を検出し、適切なアクションを実行します。
 */

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if(!getControlLayoutParent().getModifiable()){
            switch (event.getActionMasked()){
                case MotionEvent.ACTION_UP: // 1
                case MotionEvent.ACTION_POINTER_UP: // 6
                    switchButtonVisibility();
                    break;
            }
            return true;
        }

        return super.onTouchEvent(event);
    }
/**
 * 「X」の値を設定します。
 */

    @Override
    public void setX(float x) {
        super.setX(x);
        alignButtons();
    }
/**
 * 「Y」の値を設定します。
 */
    @Override
    public void setY(float y) {
        super.setY(y);
        alignButtons();
    }
/**
 * 「LayoutParams」の値を設定します。
 */
    @Override
    public void setLayoutParams(ViewGroup.LayoutParams params) {
        super.setLayoutParams(params);
        syncButtons();
    }
/**
 * 「can Snap」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public boolean canSnap(ControlInterface button) {
        boolean result = super.canSnap(button);
        return result && !containsChild(button);
    }
/**
 * 「DrawerData」の値を取得します。
 */

    //ゲッター
    public ControlDrawerData getDrawerData() {
        return drawerData;
    }
/**
 * 「load Edit Values」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public void loadEditValues(EditControlPopup editControlPopup) {
        editControlPopup.loadValues(drawerData);
    }
/**
 * 「clone Button」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public void cloneButton() {
        ControlDrawerData cloneData = new ControlDrawerData(getDrawerData());
        cloneData.properties.dynamicX = "0.5 * ${screen_width}";
        cloneData.properties.dynamicY = "0.5 * ${screen_height}";
        ((ControlLayout) getParent()).addDrawer(cloneData);
    }
/**
 * 「remove Button」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    @Override
    public void removeButton() {
        ControlLayout layout = getControlLayoutParent();
        for(ControlSubButton subButton : buttons){
            layout.removeView(subButton);
        }

        layout.getLayout().mDrawerDataList.remove(getDrawerData());
        layout.removeView(this);
    }

}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/buttons/ControlDrawer.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.buttons

import android.annotation.SuppressLint
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import com.arata.yukarilauncher.ui.subassembly.customcontrols.ControlData
import com.arata.yukarilauncher.ui.subassembly.customcontrols.ControlDrawerData
import com.arata.yukarilauncher.ui.subassembly.customcontrols.ControlLayout
import com.arata.yukarilauncher.ui.subassembly.customcontrols.handleview.EditControlPopup
import java.util.ArrayList

/**
 * コントロールドロワーのビュークラス。
 * 複数のサブボタンを格納し、指定された方向に展開表示します。
 */
@SuppressLint("ViewConstructor")
class ControlDrawer : ControlButton {
    /** このドロワーに含まれるサブボタンのリスト。 */
    val buttons: ArrayList<ControlSubButton>
    /** このドロワーのデータ。 */
    lateinit var drawerData: ControlDrawerData
    /** 親レイアウトへの参照。 */
    val parentLayout: ControlLayout
    /** サブボタンを表示するかどうか。 */
    var areButtonsVisible: Boolean
    /** ドロワーが完全に初期化されたかどうか（[drawerData]を含む）。 */
    private var mDrawerInitialized = false

    constructor(layout: ControlLayout, drawerData: ControlDrawerData) : super(layout, drawerData.properties) {
        buttons = ArrayList(drawerData.buttonProperties.size)
        parentLayout = layout
        this.drawerData = drawerData
        mDrawerInitialized = true
        areButtonsVisible = layout.isModifiable
    }

    /**
     * プロパティデータからサブボタンを追加します。
     * @param properties 追加するサブボタンのプロパティ
     */
    fun addButton(properties: ControlData) {
        addButton(ControlSubButton(parentLayout, properties, this))
    }

    /**
     * サブボタンを追加します。
     * @param button 追加するサブボタン
     */
    fun addButton(button: ControlSubButton) {
        buttons.add(button)
        syncButtons()
        setControlButtonVisibility(button, areButtonsVisible)
    }

    /**
     * サブボタンの表示状態を設定します。
     * @param button 対象のサブボタン
     * @param isVisible 表示する場合はtrue
     */
    private fun setControlButtonVisibility(button: ControlButton, isVisible: Boolean) {
        button.controlView.visibility = if (isVisible) VISIBLE else GONE
    }

    /** サブボタンの表示/非表示を切り替えます。 */
    private fun switchButtonVisibility() {
        areButtonsVisible = !areButtonsVisible
        val visibility = if (areButtonsVisible) VISIBLE else GONE
        for (button in buttons) {
            button.controlView.visibility = visibility
        }
    }

    /** サブボタンをドロワーの方向に合わせて整列します。 */
    private fun alignButtons() {
        if (!::drawerData.isInitialized) return
        if (drawerData.orientation == ControlDrawerData.Orientation.FREE) return
        val margin = ControlInterface.getMarginDistance().toInt()

        for (i in buttons.indices) {
            when (drawerData.orientation) {
                ControlDrawerData.Orientation.RIGHT -> {
                    buttons[i].setDynamicX(generateDynamicX(x + (drawerData.properties.getWidth() + margin) * (i + 1)))
                    buttons[i].setDynamicY(generateDynamicY(y))
                }
                ControlDrawerData.Orientation.LEFT -> {
                    buttons[i].setDynamicX(generateDynamicX(x - (drawerData.properties.getWidth() + margin) * (i + 1)))
                    buttons[i].setDynamicY(generateDynamicY(y))
                }
                ControlDrawerData.Orientation.UP -> {
                    buttons[i].setDynamicY(generateDynamicY(y - (drawerData.properties.getHeight() + margin) * (i + 1)))
                    buttons[i].setDynamicX(generateDynamicX(x))
                }
                ControlDrawerData.Orientation.DOWN -> {
                    buttons[i].setDynamicY(generateDynamicY(y + (drawerData.properties.getHeight() + margin) * (i + 1)))
                    buttons[i].setDynamicX(generateDynamicX(x))
                }
                else -> {}
            }
            buttons[i].updateProperties()
        }
    }

    /** サブボタンのサイズをドロワー本体に合わせます。 */
    private fun resizeButtons() {
        if (!::drawerData.isInitialized) return
        if (drawerData.orientation == ControlDrawerData.Orientation.FREE) return
        for (subButton in buttons) {
            subButton.mProperties.setWidth(mProperties.getWidth())
            subButton.mProperties.setHeight(mProperties.getHeight())
            subButton.updateProperties()
        }
    }

    /** サブボタンの整列とリサイズを同期します。 */
    fun syncButtons() {
        alignButtons()
        resizeButtons()
    }

    /**
     * 指定されたボタンがこのドロワーの子かどうか判定します。
     * @param button 判定するボタン
     * @return 子の場合はtrue
     */
    fun containsChild(button: ControlInterface): Boolean {
        for (childButton in buttons) {
            if (childButton === button) return true
        }
        return false
    }

    override fun preProcessProperties(properties: ControlData, layout: ControlLayout): ControlData {
        val data = super.preProcessProperties(properties, layout)
        data.isHideable = true
        return data
    }

    override fun setVisible(isVisible: Boolean) {
        val visibility = if (isVisible) VISIBLE else GONE
        this.visibility = visibility
        if (visibility == GONE || areButtonsVisible) {
            for (button in buttons) {
                button.controlView.visibility = if (isVisible) VISIBLE else if (!mProperties.isHideable && visibility == GONE) VISIBLE else View.GONE
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!mControlLayout.isModifiable) {
            when (event.actionMasked) {
                MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> switchButtonVisibility()
            }
            return true
        }
        return super.onTouchEvent(event)
    }

    override fun setX(x: Float) {
        super.setX(x)
        alignButtons()
    }

    override fun setY(y: Float) {
        super.setY(y)
        alignButtons()
    }

    override fun setLayoutParams(params: ViewGroup.LayoutParams) {
        super.setLayoutParams(params)
        syncButtons()
    }

    override fun canSnap(button: ControlInterface): Boolean {
        return super.canSnap(button) && !containsChild(button)
    }

    override fun loadEditValues(editControlPopup: EditControlPopup) {
        editControlPopup.loadValues(drawerData)
    }

    override fun cloneButton() {
        val cloneData = ControlDrawerData(drawerData)
        cloneData.properties.dynamicX = "0.5 * \${screen_width}"
        cloneData.properties.dynamicY = "0.5 * \${screen_height}"
        (parent as ControlLayout).addDrawer(cloneData)
    }

    override fun removeButton() {
        for (subButton in buttons) {
            parentLayout.removeView(subButton)
        }
        parentLayout.layout!!.mDrawerDataList!!.remove(drawerData)
        parentLayout.removeView(this)
    }
}
